kubeflow/trainer/rhai/utils.py

```diff
@@ -30,8 +29,52 @@ def get_trainer_cr_from_rhai_trainer(
         )
 
     else:
-        _raise_unknown_rhai_trainer(trainer)
+        raise ValueError(f"Unknown trainer {trainer}.")
+
+
+def merge_progression_annotations(
+    trainer: Union[transformers.TransformersTrainer, traininghub.TrainingHubTrainer],
+    spec_annotations: Optional[dict[str, str]] = None,
+) -> Optional[dict[str, str]]:
+    """Merge progression tracking annotations for RHAI trainers with existing annotations.
+
+    Args:
+        trainer: RHAI trainer instance (TransformersTrainer or TrainingHubTrainer).
+        spec_annotations: Existing annotations dict to merge with, if any.
+
+    Returns:
+        Merged annotations dict with progression tracking added (if enabled),
+        or original spec_annotations if progression tracking is disabled.
+    """
+    if (
+        not hasattr(trainer, "enable_progression_tracking")
+        or not trainer.enable_progression_tracking
+        or not hasattr(trainer, "metrics_port")
+        or not hasattr(trainer, "metrics_poll_interval_seconds")
+    ):
+        return spec_annotations
+
+    from kubeflow.trainer.rhai.constants import (
+        ANNOTATION_FRAMEWORK,
+        ANNOTATION_METRICS_POLL_INTERVAL,
+        ANNOTATION_METRICS_PORT,
+        ANNOTATION_PROGRESSION_TRACKING,
+    )
+
+    if isinstance(trainer, transformers.TransformersTrainer):
+        framework = "transformers"
+    elif isinstance(trainer, traininghub.TrainingHubTrainer):
+        framework = "traininghub"
+    else:
+        framework = "unknown"
 
+    progression_annotations = {
+        ANNOTATION_PROGRESSION_TRACKING: "true",
+        ANNOTATION_METRICS_PORT: str(trainer.metrics_port),
+        ANNOTATION_METRICS_POLL_INTERVAL: str(trainer.metrics_poll_interval_seconds),
+        ANNOTATION_FRAMEWORK: framework,
+    }
 
-def _raise_unknown_rhai_trainer(trainer: object) -> NoReturn:
-    raise ValueError(f"Unknown trainer {trainer}.")
+    if spec_annotations is None:
+        return progression_annotations
+    return {**spec_annotations, **progression_annotations}
```

Comment:
nit: I'd suggest merging these the other way round so users can override the progression annotations if they want. They shouldn't need to, but it may be surprising if they can't.

```suggestion
    # merge spec_annotations last to allow users to override the progress annotations
    return {**progression_annotations, **spec_annotations}
```

Reply:
here do you mean to allow user-provided annotations should have precedence ?